Slab support in data store

bvanessen opened the pull request:
Added support for placing a slab of a sample's data into the data
store.  Then each partition of data (slab) is manged as if it were a
subsample and exchanges within a fixed offset of nodes.  Also added
support in the data store to record the sample ID and slab ID.

naoyam committed:
Add environment variable LBANN_NUM_IO_PARTITIONS

Specify the number of partitions in the depth dimension of the Cosmoflow
samples.

naoyam committed:
Adjust the base offset for parallel sample I/O

naoyam committed:
WIP: Further adjustment of sample sizes

naoyam committed:
WIP: sample size adjustment

naoyam committed:
Merge branch 'distconv' into cosmoflow-parallel-io

naoyam committed:
WIP: sample size adjustment

naoyam committed:
Remove debug output

enmccarthy committed:
Cosmoflow parallel io (#86)

* before rebase

* updating

* updating

* small changes

* moving around where data is read in NOT DONE YET

* updated some comments and some todo

* cleaning up

* added comm member variabe

* cleaning up

* compiles, fixes stray variables and typos, adds correct member variables

* changing responses to float,taking away division, removing from image_data_reader

* fixing a mistake

* oops, changing m_all_responses back to float

* changed some variable names, changed indenting and fixed vim problems, fixed file access

* fixed duplicate count

* transposed dimensions, reverted resnet, took odd spacing and print statements out

* removed timing

* fixing comments and spacing

naoyam committed:
Missing semicolon

naoyam committed:
Fix type mismatch

naoyam committed:
Remove trailing whitespaces

naoyam committed:
HDF5 bug fixes

naoyam committed:
Size adjustment fix

naoyam committed:
Refactoring

naoyam committed:
Support strided rank ordering

naoyam committed:
Fix hang in HDF5 MPI-IO

HDF5 caused hanging. Likely because a HDF5 property was created with MPI
at every fetch_datum. The property is now moved out of the function and
is only done once, so it should not hang anymore. Yet, MPI-IO is disabled for now. Should be looked into again once everything becomes working.

naoyam committed:
Disables assertion

This assertion fails when the last mini-batch is not a full one. Not
sure why it fails now and not before.

bvanessen committed:
Added support for placing a slab of a sample's data into the data
store.  Then each partition of data (slab) is manged as if it were a
subsample and exchanges within a fixed offset of nodes.  Also added
support in the data store to record the sample ID and slab ID.

naoyam committed:
Use normalized parameters in Cosmoflow

naoyam committed:
Fix copying of a non-halo-expanded host tensor to a halo-extended device
tensor.

The distconv::Copy function doesn't seem to be working correctly, though
more comprehensive investigation is needed.

Commit:
Merge branch 'cosmoflow-parallel-io' of github.com:naoyam/lbann into cosmoflow-parallel-io

naoyam commented:
I don't see any issue with the changes. Let me try running this to see if the accuracy is not impacted.

bvanessen commented:
Great.  I am trying to build on rzansel right now and still moving data over.  Waiting for an allocation on lassen.

Brian C. Van Essen
vanessen1@llnl.gov
(w) 925-422-9300
(c) 925-290-5470





> On Oct 13, 2019, at 7:13 PM, Naoya Maruyama <notifications@github.com> wrote:
> 
> I don't see any issue with the changes. Let me try running this to see if the accuracy is not impacted.
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub, or unsubscribe.
>

naoyam commented on src/data_readers/data_reader_hdf5.cpp:
```diff
@@ -86,13 +124,73 @@ namespace lbann {
 
   void hdf5_reader::load() {
     lbann_comm* l_comm = get_comm();
+    if(dc::get_rank_stride() != 1) {
+      LBANN_ERROR("HDF5 MPI-IO data reader requires DistConv rank stride = 1");
```
Is there any reason with this requirement? I've been testing it with `rank_stride` == `num_partitions`.

bvanessen committed:
Testing alternate configurations of world rank.

Commit:
Fixed a performance bug where a conduit node was being copied instead
of getting an external reference.

Commit:
Merge branch 'cosmoflow-parallel-io' of github.com:bvanessen/lbann into cosmoflow-parallel-io

bvanessen committed:
Fixed a second performance bug where a new conduit node was being
passed in to the function reading the hyperslab of data, and then it
was copied outside.  Also reverted the calculation of world rank to
use the distconv methods.

bvanessen committed:
Fixed a bug where the H5Dread command was not passed the data transfer
properties.  Switched to using H5FD_MPIO_INDEPENDENT instead of
H5FD_MPIO_COLLECTIVE as the default.

naoyam commented:
This is no longer necessary as the commits are adopted in #93.
